### PR TITLE
Adds jax-cosmo implementation of challenge metrics

### PR DESCRIPTION
**Pull request by EiffL**
This PR adds the option to use [jax-cosmo](https://github.com/DifferentiableUniverseInitiative/jax_cosmo) instead of CCL/Firecrown to compute the metrics of the challenge. The benefit from this switch is that the FoM can reliably be computed, and we can even compute the DETF FoM which is unavaible in the current firecrown metrics code due to some CCL problems.

Concretely, in the config file for a run I've added this option:
```python
# Backend implementing the metrics, either: "firecrown" (default), "jax-cosmo"
metrics_impl: jax-cosmo
```
which allows the user to switch between the two, to check that I'm not gaming the metrics ^^'

I have also added a small notebook that compares the Firecrown and jax-cosmo metrics, just to demonstrate that things are working ok. Check it out here: https://github.com/LSSTDESC/tomo_challenge/blob/jax_cosmo/notebooks/jax-cosmo_validation.ipynb

### CODE REVIEW

**Comment by dkirkby** [comment]
jax_cosmo is awesome!

I am finding that I need ~50Gb of memory to calculate the (single-precision) cl's for the random forest example with 4 bins (so len(mu) = 36 x 100).   My call to `jax_cosmo.angular_cl.angular_cl` takes ~7 mins on my 32Gb laptop and a second call takes about the same amount of time and memory, so it isn't mostly initial jit compilation overhead.  Is that expected?

It would be convenient if `import tomo_challlenge.jax_metrics` didn't require that pyccl, sacc, firecrown, etc are installed.  Similarly, I guess that`import tomo_challenge.metrics` now requires that jax, xla, etc are installed.  Perhaps remove the metric imports from the top-level `__init__.py`?

**Comment by EiffL** [comment]
Thanks @dkirkby for checking it out! Yes so, it shouldn't reach 50GB of memory ^^', and 7 min seems a bit long. 

2 considerations that may explain this:
  - First, I initially wrote the metrics code to work on small batches of galaxies, for the purpose of training by binning neural network. What is does is that it builds a "differentiable n(z)" using a KDE, and therefore keeps track in memory of all the input galaxies. This is needed when backpropagating to adjust the NN weights, as the NN acts on each galaxy. So that takes some memory, but is not needed when simply evaluating the final metric. I can fix it by adding a new type of N(z) that doesn't keep track of all galaxies.

  - Jax is still pretty young, so is jax-cosmo ^^' There have been a few times where the XLA compilation of some part of the computation on CPU has been broken, but they fix it pretty fast with new Jax releases. Concretely, it took way more time to JIT on my laptop than on my desktop with GPU. This depends on the strategy used to compute parts of the cosmo equations. I've recently switched from simple linear interpolations to cubic splines, and I think this might be causing some issues in the gradient computations on CPU. Haven't had a lot of users yet ^^' I'm gonna try to reproduce your issues and maybe downgrade back to linear interpolations for now.
Also, the way jax recognizes an already jitted function is not always great, sometimes, it's only at the third evaluation that it doesnt recompile

I've been able to run 3x2pt with 5 bins without problems (on GPU ), but havent tried to push much further than that. The covariance matrix gets pretty big.... And it has to fit in memory, along with its inverse, and the gradients of both...

**Comment by EiffL** [comment]
I've made small changes.... I hope this will work better now, but you should grab the latest version of `jax-cosmo` v0.1rc5 Let me know if it works better if you get a chance :-)

**Comment by dkirkby** [comment]
Thanks @EiffL!  With your changes, the cl calculation now fits in memory and runs in about 4s (compared with 7 mins earlier).   Using fewer KDE bins doesn't speed it up further, so it looks like the nz evaluations are no longer the tall pole.

> The covariance matrix gets pretty big.... And it has to fit in memory, along with its inverse, and the gradients of both...

I just implemented sparse versions of the necessary covariance operations, taking advantage of their diagonal block structure, to learn a bit more about jax.  This reduces reduces the memory footprint by a factor of nell (100 in this example) and the SNR calculation is sped up by a similar factor:
```
cov_sparse = gaussian_cl_cov_sparse(ell, probes, mu, noise_cl, f_sky=0.25)
print(cov.size, cov_sparse.size)
12960000 129600

assert np.allclose(snr2(mu, cov), snr2_sparse(mu, cov_sparse))

%timeit snr2(mu.reshape(-1), cov).block_until_ready()
454 ms ± 23.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%timeit snr2_sparse(mu.reshape(-1), cov_sparse).block_until_ready()
4.85 ms ± 203 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
I can submit a PR to jax_cosmo with the sparse ops if that's useful.

**Comment by EiffL** [comment]
Fantastic!!!! This is incredibly awesome! With that we should be able to run batched HMC on GPU :-D I have been having memory problems preventing that so far. 
Please feel free to open a PR on `jax_cosmo` and/or on the `jax-cosmo` branch of this repo if it's a standalone piece code!

**Comment by dkirkby** [comment]
I started DifferentiableUniverseInitiative/jax_cosmo#58 with support for sparse covariances.

**Comment by EiffL** [comment]
I have updated the metric code in this branch to use the new sparse matrices. Everything looks good!

File: tomo_challenge/jax_metrics.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        # sqrt(sum(mu**2/sigma**2))
+        P = jc.sparse.inv(C)
+        score = (mu.T @ jc.sparse.sparse_dot_vec(P, mu))**0.5
+        return score
+
+    return snr_fn(probes, ell)
+
+def compute_fom(weights, labels, inds=[0,4], what='3x2', binned_nz=False):
+    """
+    Computes the omega_c, sigma8 Figure of Merit
```
Comment (joezuntz) [inline tomo_challenge/jax_metrics.py:125]: Can JAX do the w0-wa one?  I always wanted that one, it was just a limitation of CCL that we used this one.

Comment (EiffL) [reply tomo_challenge/jax_metrics.py:125]: Yes it can :-D As a matter of fact, if you specify "FOM_DETF_3x2" as the metric in the yaml file it will compute it

File: requirements.txt [hunk 2]
```diff
@@ -10,3 +10,4 @@ pyccl
 camb
 click
 progressbar
+jax-cosmo==0.1rc6
```
[the remark this answers is not in the document]
Comment (EiffL) [reply requirements.txt:13]: yep, no pain, all gain.  That's also part of the mission statement of  `jax-cosmo` it should be trivially pip installable.

File: tomo_challenge/jax_metrics.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
+import jax.numpy as np
```
[the remark this answers is not in the document]
Comment (EiffL) [reply tomo_challenge/jax_metrics.py:1]: hummm good point..... I'll hide this, the same way I've hidden firecorwn and ccl imports

**Comment by EiffL** [comment]
Alrighty then, I'm merging :-)
